File: core/documents/tutorial_prerequisite_engine.py
from __future__ import annotations

from typing import Any, Dict, List

from core.documents.instructional_semantics_engine import analyze_instructional_semantics
from core.documents.tutorial_dependency_engine import reconstruct_tutorial_dependencies


def infer_tutorial_prerequisites(text: str) -> Dict[str, Any]:
    inst = analyze_instructional_semantics(text)
    legacy = reconstruct_tutorial_dependencies(text)
    chain: List[Dict[str, str]] = []
    steps = inst.get("ordering", [])
    for i in range(1, len(steps)):
        chain.append({
            "prerequisite": steps[i - 1].get("title", ""),
            "requires": steps[i].get("title", ""),
            "evidence": "discourse:instructional_order",
        })
    return {
        "chain": chain,
        "prerequisites": inst.get("prerequisites", []),
        "legacy_flow": legacy.get("reconciled", {}),
        "deterministic_inputs": [f"chain={len(chain)}"],
    }

File: src/documents/tutorialPrerequisiteEngine.ts
/**
 * Converted from Python: core/documents/tutorial_prerequisite_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { analyzeInstructionalSemantics } from "./instructionalSemanticsEngine.js";
import { reconstructTutorialDependencies } from "./tutorialDependencyEngine.js";

export function inferTutorialPrerequisites(text: any): any {
  var inst: any = analyzeInstructionalSemantics(text);
  var legacy: any = reconstructTutorialDependencies(text);
  var chain: any[] = [];
  var steps: any = py.get(inst, "ordering", []);
  var i: any;
  for (i = 1; i < py.len(steps); i++) {
    py.listAppend(chain, {"prerequisite": py.get(py.at(steps, py.sub(i, 1)), "title", ""), "requires": py.get(py.at(steps, i), "title", ""), "evidence": "discourse:instructional_order"});
  }
  return {"chain": chain, "prerequisites": py.get(inst, "prerequisites", []), "legacy_flow": py.get(legacy, "reconciled", {}), "deterministic_inputs": [`chain=${py.toStr(py.len(chain))}`]};
}
export { analyzeInstructionalSemantics, reconstructTutorialDependencies };
